File: LivrusVirtus/FoxToolKit/FXPython/RotinasGlobais.py
from FXPy.fox import *
from time import localtime, strftime 
#, strptime não existe na versão 2.2 do Python
from resource import Resource
from ListaStr import LISTAMSTR
import fpformat
import os

def create():
  return RotinasGlobais()

class RotinasGlobais:
  gConexao = None
  sBarraLogin = ''
  sUsuarioLogin = ''
  sUsuarioSenha = ''
  sCodigoSelecionado = ''
  sPreco = ''
  sQtdEstoque = ''
  sepDTHR = '#' #SOMENTE ACCESS
  #sepDTHR = "'" #OUTROS BANCOS

  def __init__(self):
    self.data = []

  def lerFigura(self, form, nomearquivo):
    return FXBMPIcon(form, 
      open(os.path.join('bmps', nomearquivo), 'rb').read())

  def ValidaFloat(self, sNum):
    nNum = 0.0
    bret = False

    if len(sNum) > 0:
      try:
        nNum = float(sNum)
        if nNum >= 0.01:
          bret = True
      except:
        showModalErrorBox(self, MBOX_OK, 'Erro', Resource.STR_NUMINV)
        bret = False
    else:
      bret = False
    
    return bret
  
  def VirgulaParaPonto(self, sNum, bRevertido):
    if bRevertido:
      sret = sNum.replace('.', ',')
    else:
      sret = sNum.replace(',', '.')
    
    return sret

  def substr(self, stexto, nini, ntam):
    return stexto[nini-1:nini+ntam-1]

  def modulo11(self, svalor, ndig):
    ncont = 1
    slocal = ''
    while (ncont <= 11):
      slocal=self.substr(svalor,ncont,1) + slocal
      ncont=ncont+1
    nsoma=0;
    if (ndig == 1):
      ncont=3;
    else:
      ncont=2;
    npos = 2
    while (ncont <= 11):
      sres = self.substr(slocal,ncont,1)
      nvalor = int(sres)
      nsoma=nsoma + (nvalor * npos)
      ncont=ncont+1
      npos=npos+1
  
    nsoma=(11 - (nsoma % 11))
    if (nsoma > 9):
      return 0
    else:
      return nsoma

  def validaCPF(self, cpf):
    if len(cpf)==14:
      try:
        scpf=self.substr(cpf,1,3)+self.substr(cpf,5,3)+self.substr(cpf,9,3)+\
          self.substr(cpf,13,2)
        sres=self.substr(cpf,13,1)
        nvalor1 = int(sres)
        sres=self.substr(cpf,14,1)
        nvalor2 = int(sres)
        if (self.modulo11(scpf,1)==nvalor1) and\
           (self.modulo11(scpf,2)==nvalor2):
          return True
        else:
          return False
      except:
        return False
    else:
      return False

  def validaISBN(self, isbn):
    if len(isbn)==13:
      sdigito = self.substr(isbn, len(isbn), 1)
      sisbn=''
      for ncont in range(1, len(isbn)):
        if self.substr(isbn,ncont,1)!='-':
          sisbn=sisbn+self.substr(isbn,ncont,1)
  
      if sdigito == 'X':
        nsoma = 10
      else:
        nvalor = int(sdigito)
        nsoma = nvalor
  
      for ncont in range(1, 10):
        snum = self.substr(sisbn, ncont, 1)
        nvalor = int(snum)
        nsoma=nsoma + (nvalor * (11 - ncont))
  
      if ((nsoma % 11) == 0):
        return True
      else:
        return False
    else:
      return False

  def validaDataHora(self, sDataHora, bSoData):
    bvalida=False
    bdatavalida=False
    bhoravalida=False
    nudiames=0
    if bSoData==True: ntam=10
    else: ntam=19

    if len(sDataHora)==ntam:
      sdia = self.substr(sDataHora,1,2)
      ndia = int(sdia)
      smes = self.substr(sDataHora,4,2)
      nmes = int(smes)
      sano = self.substr(sDataHora,7,4)
      nano = int(sano)
    
      if (nano >= 1) and (nano <= 9999):
        if (nmes >= 1) and (nmes <= 31):
            if nmes in [1, 3, 5, 7, 8, 10, 12]: 
              nudiames = 31
            elif nmes == 2: 
              # ano bissexto 
              if (nano % 4) == 0: 
                nudiames = 29
              else: nudiames = 28
            elif nmes in [4, 6, 9, 11]: 
              nudiames = 30
            if (ndia >=1) and (ndia <=nudiames):
               bdatavalida = True
    
      if bSoData:
         if bdatavalida: bvalida = True
         else: bvalida = False
      else:
        if bdatavalida:
            shora = self.substr(sDataHora,12,2)
            nhora = int(shora)
            smin = self.substr(sDataHora,15,2)
            nmin = int(smin)
            sseg = self.substr(sDataHora,18,2)
            nseg = int(sseg)
        
            if (nhora >= 0) and (nhora <= 23):
              if (nmin >= 0) and (nmin <= 59):
                if (nseg >= 0) and (nseg <= 59):
                   bhoravalida = True
        
            if bhoravalida: bvalida = True
            else: bvalida = False
        else:
          bvalida = False
    else:
      bvalida = False
  
    return bvalida

  def retDataHoraStr(self, bData, bHora):
    sDataHora = ''
    
    if bData and bHora:
      sDataHora = strftime('%d/%m/%Y %H:%M:%S', localtime())
    elif bData:
      sDataHora = strftime('%d/%m/%Y', localtime())
    elif bHora:
      sDataHora = strftime('%H:%M:%S', localtime())
    
    return sDataHora;

  def ConverteDataHoraStr(self, sDTHR, bDMY_ou_YMD, sFrmtData, sFrmtHora):
    sDataHora = ''
    sFormato = ''
    nano, nmes, ndia, nhora, nmin, nseg = 2005, 1, 1, 0, 0, 0

    if len(sFrmtData) > 0:
      if bDMY_ou_YMD:
        ndia = int(self.substr(sDTHR, 1, 2))
        nmes = int(self.substr(sDTHR, 4, 2))
        nano = int(self.substr(sDTHR, 7, 4))
      else:
        nano = int(self.substr(sDTHR, 1, 4))
        nmes = int(self.substr(sDTHR, 6, 2))
        ndia = int(self.substr(sDTHR, 9, 2))
      #sDataHora = ndia + '/' + nmes + '/' + nano
      #sFormato = '%d/%m/%Y'

    if len(sFrmtHora) > 0:
      nhora = int(self.substr(sDTHR, 12, 2))
      nmin = int(self.substr(sDTHR, 15, 2))
      nseg = int(self.substr(sDTHR, 18, 2))
      #sDataHora = sDataHora + nhora + ':' + nmin + ':' + nseg
      #sFormato = sFormato + '%H:%M:%S'

    #sDTHR = strptime(sDataHora, sFormato)
    sDTHR = (nano, nmes, ndia, nhora, nmin, nseg, 0, 1, 0)
    sDataHora = strftime( sFrmtData + ' ' + sFrmtHora , sDTHR)
    return sDataHora

  def ConectaBanco(self, sBancoDeDados, sLogin, sSenha):
    import odbc
    #import MySQLdb
    try:
      #self.gConexao = MySQLdb.connect( host = 'localhost', 
      #  db = sBancoDeDados, user = sLogin, passwd = sSenha )
      self.gConexao = odbc.odbc( sBancoDeDados )
    except:
      showModalErrorBox(self, MBOX_OK, 'Erro', Resource.STR_NCONBANCO)
      return False
    else:
      return True
  
  def DisconectaBanco(self):
    pass
  
  def AtualizaDados(self, hBancoLocal, sSql):
    try:
      cursor = hBancoLocal.cursor()
      # Necessario codificacao de acordo com o banco
      cursor.execute( sSql.encode('latin-1') )   
    except:
      showModalErrorBox(self, MBOX_OK, 'Erro', Resource.STR_NCEXECSQL)
      return False
    else:
      return True
  
  def ConsultaDados(self, hBancoLocal, sSql):
    try:
      cursor = hBancoLocal.cursor()
      cursor.execute( sSql )   
    except:
      showModalErrorBox(self, MBOX_OK, 'Erro', Resource.STR_NCEXECSQL)
      return False
    else:
      LISTAMSTR.col = cursor.fetchall()
      LISTAMSTR.tam = len(LISTAMSTR.col)
      if LISTAMSTR.tam > 0:
        return True
      else:
        return False

  def strptimeFixo(self, sDataHora):
    lmes = {'Jan':1,'Feb':2, 'Mar':3 , 'Apr':4, 'May':5, 'Jun':6
    , 'Jul':7, 'Aug':8, 'Sep':9, 'Oct':10, 'Nov':11, 'Dec':12 }
    #nwdia = self.substr(sDataHora,1,3)		
    nmes = lmes[self.substr(sDataHora,5,3)]
    ndia = int(self.substr(sDataHora,9,2))
    nhora = int(self.substr(sDataHora,12,2))
    nmin = int(self.substr(sDataHora,15,2))
    nseg = int(self.substr(sDataHora,18,2))
    nano = int(self.substr(sDataHora,21,4))
  
    sDTHR = (nano, nmes, ndia, nhora, nmin, nseg, 0, 1, 0)
    return sDTHR

  def arrumaColunas(self, lstLocal):
    lstLinhas = []
    nQtdCols = len(lstLocal[0])
    for nLinha in range(0,len(lstLocal)):
      lstCols = []
      for nCol in range(0,nQtdCols):
        if type(lstLocal[nLinha][nCol]) is str:
          lstCols.append(lstLocal[nLinha][nCol]) 
        elif type(lstLocal[nLinha][nCol]) is int:
          lstCols.append(str(lstLocal[nLinha][nCol])) 
        elif type(lstLocal[nLinha][nCol]) is float:
          lstCols.append(fpformat.fix(lstLocal[nLinha][nCol],2)) 
        else:
          #sDTHR = strptime( str(lstLocal[nLinha][nCol]), "%a %b %d %H:%M:%S %Y")
          sDTHR = self.strptimeFixo( str(lstLocal[nLinha][nCol]) )
          lstCols.append(strftime( '%d/%m/%Y %H:%M:%S', sDTHR)) 
      
      lstLinhas.append(lstCols) 
  
    return lstLinhas

  def ConsultaDadosLista(self, hBancoLocal, lstLocal, sSql):
    try:
      cursor = hBancoLocal.cursor()
      cursor.execute( sSql )   
    except:
      showModalErrorBox(self, MBOX_OK, 'Erro', Resource.STR_NCEXECSQL)
      return False
    else:
      lstLocal.col = cursor.fetchall()
      lstLocal.tam = len(lstLocal.col)
      if lstLocal.tam > 0:
        lstLocal.col = self.arrumaColunas(lstLocal.col)
        return True
      else:
        return False

  def ValidaCampo(self, sCampo, sMensagemErro):
    bret = False
    if len(sCampo)==0:
      showModalErrorBox(self, MBOX_OK, 'Erro', sMensagemErro)
      bret = True
    else:
  	  bret = False
  
    return bret

  def ConsultaCampoDesc(self, Campos, sTextoSql, sMensagemErro):
    sret = ""
    if self.ConsultaDadosLista(self.gConexao, Campos, sTextoSql):
      sret = Campos.col[0][1]
    else:
      showModalErrorBox(self, MBOX_OK, 'Erro', sMensagemErro)
      sret = ""
    
    return sret

  def AdicionaItemLista(self, lstLista, slCodLista, sCampoCodigo, sCampoDesc):
    if len(sCampoDesc.getText()) > 0:
      sCodigo = sCampoCodigo.getText()
      sDesc = sCampoDesc.getText()
      
      sItems = sCodigo + ' - ' + sDesc
      
      if lstLista.findItem(sItems) == -1:
        slCodLista.col.append(sCodigo) 
        lstLista.appendItem(sItems)
      sCampoCodigo.setText("")
      sCampoDesc.setText("")
  
  def RemoveItemLista(self, lstLista, slCodLista):
    if lstLista.getNumItems() > 0:
      nSelIndice = lstLista.getCurrentItem()
      if lstLista.isItemCurrent(nSelIndice):
        slCodLista.col.pop(nSelIndice)
        lstLista.removeItem(nSelIndice)
  
  def AdicionaValoresLista(self, sTextoSql, lstLista, slCodLista):
    lstCampos = LISTAMSTR()
    nLinha = 0
  
    lstLista.clearItems()
    slCodLista.col = []
    if self.ConsultaDadosLista(self.gConexao, lstCampos, sTextoSql):
      for nLinha in range(0,lstCampos.tam):
        sCodigo = lstCampos.col[nLinha][0]
        sDesc = lstCampos.col[nLinha][1]
  
        slCodLista.col.append(lstCampos.col[nLinha][0])
        sItems = sCodigo + ' - ' + sDesc
        lstLista.appendItem(sItems)
  
  def AdicionaColunasGrid(self, grid, nPosIni, nQtdCols, nLargura, sRotulos):
    #if (nPosIni > 1):
    #  grid.insertColumns(nPosIni, nQtdCols)
    nCont = 0
    for nPos in range(nPosIni,(nPosIni+nQtdCols)):
      grid.setColumnWidth(nPos, nLargura[nCont])
      grid.setItemText(0, nPos, sRotulos[nCont])  
      nCont = nCont + 1
  
  def AdicionaLinhasGrid(self, grid, lstLinhas, nQtdCol):
    if grid.getNumRows() > 1:
      grid.removeRows(1, grid.getNumRows()-1)  
    grid.insertRows(1, lstLinhas.tam)

    for nLinha in range(1, grid.getNumRows()):
      grid.setItemText(nLinha, 0, str(nLinha))
      grid.getItem(nLinha, 0).setButton(True)

    for nCol in range(1, nQtdCol+1):
      for nLinha in range(1, lstLinhas.tam+1):
        sLinha = lstLinhas.col[nLinha-1][nCol-1]
        grid.setItemText(nLinha, nCol, sLinha)
        if nCol > 1:
          grid.getItem(nLinha, nCol).justify = FXTableItem.LEFT

  def IniciaLayoutGrid(self, grid, nQtdLinhas, nQtdCols):
    grid.setTableSize(nQtdLinhas, nQtdCols)

    grid.leadingRows = 1
    grid.leadingCols = 1
    #grid.trailingRows = 1
    #grid.trailingCols = 1

    grid.setItemText( 0,  0, "")
    grid.setItemText( 0, nQtdCols-1, "")
    grid.setItemText(nQtdLinhas-1,  0, "")
    grid.setItemText(nQtdLinhas-1, nQtdCols-1, "")
    
    for nCol in range(1, nQtdCols):
      grid.setItemText(0, nCol, "")
      grid.getItem(0, nCol).setButton(True)
      grid.getItem(0, nCol).justify = FXTableItem.LEFT
    
    for nLinha in range(1, nQtdLinhas):
      grid.setItemText(nLinha, 0, str(nLinha))
      grid.getItem(nLinha, 0).setButton(True)

    grid.getItem(0, 0).setButton(True)

    grid.setItemText(0, 1, "Código")
    grid.setColumnWidth(1, 80)
    grid.setItemText(0, 2, "Descrição")
    grid.setColumnWidth(2, 180)

File: LivrusVirtus/FoxToolKit/FXRuby/RotinasGlobais.rb
require 'fox'
require 'DBI'
# require 'DBD::ODBC'

require 'resource'
require 'ListaStr'

include Fox
$Rotinas = nil

class RotinasGlobais
  attr :gConexao, true
  attr :sBarraLogin, true
  attr :sUsuarioLogin, true
  attr :sUsuarioSenha, true
  attr :sCodigoSelecionado, true
  attr :sPreco, true
  attr :sQtdEstoque, true
  SepDTHR = "#" #SOMENTE ACCESS
  #SepDTHR = "'" #OUTROS BANCOS

  # Ler uma figura do arquivo
  def lerFigura(form, nomearquivo)
    begin
      nomearquivo = File.join("bmps", nomearquivo)
      figura = nil
      File.open(nomearquivo, "rb") { |f|
        #figura = FXPNGIcon.new(getApp(), f.read)
	      figura = FXBMPIcon.new(form, f.read)
      }
      return figura
    rescue
      raise RuntimeError, "Não posso ler a figura: #{nomearquivo}"
    end
  end

  def ValidaFloat(sNum)
    if sNum != ""
      if sNum.to_f > 0
        return true
      else
        FXMessageBox.error(nil, MBOX_OK, "Erro", Resource::STR_NUMINV)
        return false
      end
    else
     return false
    end
  end 

  def VirgulaParaPonto(sNum, bRevertido)
    if (bRevertido)
      return sNum.sub(/[.]/, ',')
    else
      return sNum.sub(/[,]/, '.')
    end
  end

  def modulo11(svalor, ndig)
    ncont, nsoma, npos, nvalor = 0,0,0,0
    sres, slocal = "", ""
  
    ncont=1
    while ncont <= 11 do
      slocal=svalor.slice(ncont-1,1) + slocal
      ncont=ncont+1
    end
    nsoma=0
    if ndig == 1 then
      ncont=3
    else
      ncont=2
    end
    npos=2
    while ncont <= 11 do
      sres=slocal.slice(ncont-1,1)
      nvalor = sres.to_i
      nsoma=nsoma + (nvalor * npos)
      ncont=ncont+1
      npos=npos+1
    end
    nsoma=(11 - (nsoma % 11))
    
    if nsoma > 9
      return 0
    else
      return nsoma
    end
  end

  def validaCPF(cpf)
    scpf, sres = "", ""
    nvalor1, nvalor2 = 0,0
    
    if cpf.length==14 then
      scpf=cpf.slice(0,3)+cpf.slice(4,3)+cpf.slice(8,3)+cpf.slice(12,2)
      sres=cpf.slice(12,1)
      nvalor1 = sres.to_i
      sres=cpf.slice(13,1)
      nvalor2 = sres.to_i
      if (modulo11(scpf,1)==nvalor1) and (modulo11(scpf,2)==nvalor2)
        return true
      else
        return false
      end
    else
      return false
    end
  end

  def validaISBN(isbn)
    snum, sdigito, sisbn = "", "", ""
    ncont, nsoma, nvalor = 0,0,0
  
    if isbn.length==13 then
      sdigito=isbn.slice(isbn.length-1, 1)
      sisbn = ""
      for ncont in (1..isbn.length-1) do
        if isbn.slice(ncont-1,1)!="-" then
          sisbn=sisbn+isbn.slice(ncont-1,1)
        end
      end
      if sdigito == "X" then
        nsoma = 10
      else
        nvalor = sdigito.to_i
        nsoma = nvalor
      end
      for ncont in (1..9) do
        snum=sisbn.slice(ncont-1, 1)
        nvalor = snum.to_i
        nsoma=nsoma + (nvalor * (11 - ncont))
      end
      if (nsoma % 11) == 0
        return true
      else
        return false
      end
    else
      return false
    end
  end

  def validaDataHora(sDataHora, bSoData)
    sdia, smes, sano, shora, smin, sseg = "","","","","",""
    ndia, nmes, nano, nhora, nmin, nseg, ntam = 0,0,0,0,0,0,0 
    
    bvalida=false
    bdatavalida=false
    bhoravalida=false
    nudiames=0
    
    if bSoData==true then ntam=10
    else ntam=19
    end
    if sDataHora.length==ntam then
      sdia=sDataHora.slice(0,2)
      ndia=sdia.to_i
      smes=sDataHora.slice(3,2)
      nmes=smes.to_i
      sano=sDataHora.slice(6,4)
      nano=sano.to_i
    
      if (nano >= 1) and (nano <= 9999) then
        if (nmes >= 1) and (nmes <= 31) then
            case nmes
              when 1, 3, 5, 7, 8, 10, 12
                nudiames=31
              when 2
                 # ano bissexto
                 if (nano % 4)==0 then 
                   nudiames=29
                 else 
                   nudiames=28
                 end
               when 4, 6, 9, 11
                 nudiames=30
            end
            if (ndia >=1) and (ndia <=nudiames) then
               bdatavalida=true
            end 
        end
      end
    
      if bSoData then
         if bdatavalida then 
           bvalida=true
         else 
           bvalida=false
         end 
      else
        if bdatavalida then
            shora=sDataHora.slice(11,2)
            nhora = shora.to_i
            smin=sDataHora.slice(14,2)
            nmin = smin.to_i
            sseg=sDataHora.slice(17,2)
            nseg = sseg.to_i
    
            if (nhora >= 0) and (nhora <= 23) then
              if (nmin >= 0) and (nmin <= 59) then
                if (nseg >= 0) and (nseg <= 59) then
                   bhoravalida=true
                end
              end
            end
    
            if bhoravalida then 
               bvalida=true
            else 
               bvalida=false
            end
        else
          bvalida=false
        end
      end
    else
      bvalida=false
    end
    return bvalida
  end

  def retDataHoraStr(bData, bHora)
    sDataHora = ''
    sTempo = Time.now 
    if (bData) and (bHora) then
      sDataHora = sTempo.strftime("%d/%m/%Y %H:%M:%S")
    elsif bData then
      sDataHora = sTempo.strftime("%d/%m/%Y")
    elsif bHora then
      sDataHora = sTempo.strftime("%H:%M:%S")
    end
    return sDataHora
  end

  def ConverteDataHoraStr(sDTHR, bDMY_ou_YMD, sFrmtData, sFrmtHora)
    sDataHora = ''
    
    if sFrmtData.length > 0 then
      if bDMY_ou_YMD then
        ndia = sDTHR.slice(0, 2)
        nmes = sDTHR.slice(3, 2)
        nano = sDTHR.slice(6, 4)
      else
        nano = sDTHR.slice(0, 4)
        nmes = sDTHR.slice(5, 2)
        ndia = sDTHR.slice(8, 2)
      end
    end
    if sFrmtHora.length > 0 then
      nhora = sDTHR.slice(11, 2)
      nmin = sDTHR.slice(14, 2)
      nseg = sDTHR.slice(17, 2)
    end
    sTempo = Time.local(nano, nmes, ndia, nhora, nmin, nseg, 0)
    if sFrmtHora.length > 0 then
      sDataHora = sTempo.strftime( sFrmtData + ' ' + sFrmtHora )
    else
      sDataHora = sTempo.strftime( sFrmtData + ' 00:00:00' )
    end
    return sDataHora
  end

  def ConectaBanco(sBancoDeDados, sLogin, sSenha)
    begin
      self.gConexao = DBI.connect( sBancoDeDados, sLogin, sSenha ) 
      return true
    rescue
      FXMessageBox.error(nil, MBOX_OK, "Erro", Resource::STR_NCONBANCO)
      return false
    end
  end
  
  def DisconectaBanco()
    self.gConexao.disconnect()
  end

  def AtualizaDados(hBancoLocal, sSql)
    begin
      sth = hBancoLocal.prepare( sSql )
      sth.execute()
      sth.finish()
      return true
    rescue
      FXMessageBox.error(nil, MBOX_OK, "Erro", Resource::STR_NCEXECSQL)
      return false
    end
  end

  def ConsultaDados(hBancoLocal, sSql)
    sth = hBancoLocal.prepare( sSql )
    rc = sth.execute()
    if rc then
      $ListaMStr.col = sth.fetch
      if $ListaMStr.col then
        $ListaMStr.tam = $ListaMStr.col.length
        sth.finish()
        if $ListaMStr.tam > 0 then
          return true
        else
          return false
        end
      else
        return false
      end
    else
      FXMessageBox.error(nil, MBOX_OK, "Erro", Resource::STR_NCEXECSQL)
      sth.finish()
      return false
    end
  end

  def arrumaColunas(sth)
    lstLinhas = []
    nCol = 0
    nLinha = 0
    while sDBILinha = sth.fetch do
      lstCols = []		
      for nCol in (0..sDBILinha.length-1) do
        if sDBILinha[nCol].class == String then
          lstCols[nCol] = sDBILinha[nCol]
        elsif sDBILinha[nCol].class == Fixnum then
          lstCols[nCol] = sDBILinha[nCol].to_s
        elsif sDBILinha[nCol].class == Float then
          lstCols[nCol] = "%.2f" % sDBILinha[nCol]
        else
          lstCols[nCol] = sDBILinha[nCol].to_time.strftime('%d/%m/%Y %H:%M:%S')
        end
      end
      lstLinhas[nLinha] = lstCols
      nLinha+=1
    end
    return lstLinhas
  end

  def ConsultaDadosLista(hBancoLocal, lstLocal, sSql)
    sth = hBancoLocal.prepare( sSql )
    rc = sth.execute()
    if rc then
      lstLocal.col = arrumaColunas(sth)
      if lstLocal.col then
        lstLocal.tam = lstLocal.col.length     
        if lstLocal.tam > 0 then
          sth.finish()
          return true
        end
      else
        return false
      end
    else
      FXMessageBox.error(nil, MBOX_OK, "Erro", Resource::STR_NCEXECSQL)
      sth.finish()
      return false
    end
  end

  def ValidaCampo(sCampo, sMensagemErro)
    if sCampo.strip == "" then
      FXMessageBox.error(nil, MBOX_OK, "Erro", sMensagemErro)
      return true
    else
      return false
    end
  end

  def ConsultaCampoDesc(lstCampos, sTextoSql, sMensagemErro)
    sret = ""
    if ConsultaDadosLista(self.gConexao, lstCampos, sTextoSql) then
      sret = lstCampos.col[0][1]
    else
      #FXMessageBox.error(nil, MBOX_OK, "Erro", sMensagemErro)
      sret = ""
    end
    return sret
  end

  def AdicionaItemLista(lstLista, slCodLista, sCampoCodigo, sCampoDesc)   
    if sCampoDesc.getText().length > 0    
      sCodigo = sCampoCodigo.getText()
      sDesc = sCampoDesc.getText()
      
      sItems = sCodigo + ' - ' + sDesc
      
      if lstLista.findItem(sItems) == -1
        slCodLista.col.push(sCodigo) 
        lstLista.appendItem(sItems)
      end
      sCampoCodigo.setText("")
      sCampoDesc.setText("")
    end 
  end
  
  def RemoveItemLista(lstLista, slCodLista)
     if lstLista.getNumItems() > 0
       nSelIndice = lstLista.getCurrentItem()
       if lstLista.isItemCurrent(nSelIndice)
         slCodLista.col.delete_at(nSelIndice)
    	   lstLista.removeItem(nSelIndice)
       end
     end
  end
  
  def AdicionaValoresLista(sTextoSql, lstLista, slCodLista)
    lstCampos = LISTAMSTR.new
    nLinha = 0
  
    lstLista.clearItems()
    slCodLista.col = []
    if ConsultaDadosLista(self.gConexao, lstCampos, sTextoSql)
      for nLinha in (0..lstCampos.tam-1) do
        sCodigo = lstCampos.col[nLinha][0]
        sDesc = lstCampos.col[nLinha][1]
  
  	    slCodLista.col.push(lstCampos.col[nLinha][0])
        sItems = sCodigo + ' - ' + sDesc
  	    lstLista.appendItem(sItems)
  	  end
    end
  end
  
  def AdicionaColunasGrid(grid, nPosIni, nQtdCols, nLargura, sRotulos) 
    #if (nPosIni > 1)
    #  grid.insertColumns(nPosIni, nQtdCols)
    #end
    nCont = 0
    for nPos in (nPosIni..(nPosIni+nQtdCols-1)) do
      grid.setColumnWidth(nPos, nLargura[nCont])
      grid.setItemText(0, nPos, sRotulos[nCont])  
      nCont = nCont + 1
    end
  end
  
  def AdicionaLinhasGrid(grid, lstLinhas, nQtdCol)  
    if (grid.getNumRows() > 1)
      grid.removeRows(1, grid.getNumRows()-1)  
    end  
    grid.insertRows(1, lstLinhas.tam())

    for nLinha in (1..grid.getNumRows()-1) do
      grid.setItemText(nLinha, 0, nLinha.to_s)
      grid.getItem(nLinha, 0).setButton(true)
    end
    for nCol in (1..nQtdCol) do
      for nLinha in (1..lstLinhas.tam) do
        sLinha = lstLinhas.col[nLinha-1][nCol-1]
        grid.setItemText(nLinha, nCol, sLinha)
        grid.getItem(nLinha, nCol).justify = FXTableItem::RIGHT
        if (nCol > 1)
          grid.getItem(nLinha, nCol).justify = FXTableItem::LEFT
        end
      end
    end
  end

  def IniciaLayoutGrid(grid, nQtdLinhas, nQtdCols)
    grid.setTableSize(nQtdLinhas, nQtdCols)

    grid.leadingRows = 1
    grid.leadingCols = 1
    #grid.trailingRows = 1
    #grid.trailingCols = 1

    grid.setItemText( 0,  0, "")
    grid.setItemText( 0, nQtdCols-1, "")
    grid.setItemText(nQtdLinhas-1,  0, "")
    grid.setItemText(nQtdLinhas-1, nQtdCols-1, "")
    
    for nCol in (1..nQtdCols-1) do
      grid.setItemText(0, nCol, "")
      grid.getItem(0, nCol).setButton(true)
      grid.getItem(0, nCol).justify = FXTableItem::LEFT
    end
    
    for nLinha in (1..nQtdLinhas-1) do
      grid.setItemText(nLinha, 0, nLinha.to_s)
      grid.getItem(nLinha, 0).setButton(true)
    end
    grid.getItem(0, 0).setButton(true)

    grid.setItemText(0, 1, "Código")
    grid.setColumnWidth(1, 80)
    grid.setItemText(0, 2, "Descrição")
    grid.setColumnWidth(2, 180)
  end

end
